corehq/apps/registry/models.py

```diff
@@ -124,3 +142,167 @@ class RegistryPermission(models.Model):
+        (ACTION_INVITATION_REJECTED, _("Invitation Rejected")),
+        (ACTION_GRANT_ADDED, _("Grant created")),
+        (ACTION_GRANT_REMOVED, _("Grant removed")),
+        (ACTION_SCHEMA_CHANGED, _("Schema Changed")),
+        (ACTION_DATA_ACCESSED, _("Data Accessed")),
+    )
+
+    RELATED_OBJECT_REGISTRY = "registry"
+    RELATED_OBJECT_INVITATION = "invitation"
+    RELATED_OBJECT_GRANT = "grant"
+    RELATED_OBJECT_UCR = "ucr"
+    RELATED_OBJECT_APPLICATION = "application"  # case search
+    RELATED_OBJECT_CHOICES = (
+        (RELATED_OBJECT_REGISTRY, _("Data Registry")),
+        (RELATED_OBJECT_INVITATION, _("Invitation")),
+        (RELATED_OBJECT_GRANT, _("Grant")),
+        (RELATED_OBJECT_UCR, _("Report")),
+        (RELATED_OBJECT_APPLICATION, _("Case Search")),
+    )
+
+    registry = models.ForeignKey("DataRegistry", related_name="audit_logs", on_delete=models.CASCADE)
+    date = models.DateTimeField(auto_now_add=True)
+    action = models.CharField(max_length=32, choices=ACTION_CHOICES)
+    domain = models.CharField(max_length=255)
+    user = models.ForeignKey(User, related_name="registry_actions", on_delete=models.CASCADE)
+    related_object_id = models.CharField(max_length=32)
+    related_object_type = models.CharField(max_length=32, choices=RELATED_OBJECT_CHOICES)
+    detail = JSONField(null=True)
+
+
+class RegistryAuditHelper:
+    def __init__(self, registry):
+        self.registry = registry
+
+    def registry_activated(self, user):
+        self.log_registry_activated_deactivated(user, is_activated=True)
+
+    def registry_deactivated(self, user):
+        self.log_registry_activated_deactivated(user, is_activated=False)
+
+    def invitation_accepted(self, user, invitation):
+        return self.log_invitation_accepted_rejected(user, invitation, is_accepted=True)
+
+    def invitation_rejected(self, user, invitation):
+        return self.log_invitation_accepted_rejected(user, invitation, is_accepted=False)
+
+    def invitation_added(self, user, invitation):
+        return self.log_invitation_added_removed(user, invitation, is_added=True)
+
+    def invitation_removed(self, user, invitation):
+        return self.log_invitation_added_removed(user, invitation, is_added=False)
+
+    def grant_added(self, user, grant):
+        return self.log_grant_added_removed(user, grant, is_added=True)
+
+    def grant_removed(self, user, grant):
+        return self.log_grant_added_removed(user, grant, is_added=False)
+
+    def schema_changed(self, user, new, old):
+        return RegistryAuditLog.objects.create(
+            registry=self.registry,
+            user=user,
+            action=RegistryAuditLog.ACTION_SCHEMA_CHANGED,
+            domain=self.registry.domain,
+            related_object_id=self.registry.id,
+            related_object_type=RegistryAuditLog.RELATED_OBJECT_REGISTRY,
+            detail={
+                "new_schema": new,
+                "old_schema": old,
+            }
```

Comment:
:+1: 
Maybe getting a bit ahead of things, but I could imagine this being used for an "undo" workflow, even if only a mostly-manual one done by the support team.